enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,500 @@
+			return xerrors.Errorf("failed to get template presets with prebuilds: %w", err)
+		}
+		if len(presetsWithPrebuilds) == 0 {
+			return nil
+		}
+		allRunningPrebuilds, err := db.GetRunningPrebuiltWorkspaces(ctx)
+		if err != nil {
+			return xerrors.Errorf("failed to get running prebuilds: %w", err)
+		}
+
+		allPrebuildsInProgress, err := db.CountInProgressPrebuilds(ctx)
+		if err != nil {
+			return xerrors.Errorf("failed to get prebuilds in progress: %w", err)
+		}
+
+		presetsBackoff, err := db.GetPresetsBackoff(ctx, c.clock.Now().Add(-c.cfg.ReconciliationBackoffLookback.Value()))
+		if err != nil {
+			return xerrors.Errorf("failed to get backoffs for presets: %w", err)
+		}
+
+		state = prebuilds.NewGlobalSnapshot(presetsWithPrebuilds, allRunningPrebuilds, allPrebuildsInProgress, presetsBackoff)
+		return nil
+	}, &database.TxOptions{
+		Isolation:    sql.LevelRepeatableRead, // This mirrors the MVCC snapshotting Postgres does when using CTEs
+		ReadOnly:     true,
+		TxIdentifier: "prebuilds_state_determination",
+	})
+
+	return &state, err
+}
+
+func (c *StoreReconciler) ReconcilePreset(ctx context.Context, ps prebuilds.PresetSnapshot) error {
+	logger := c.logger.With(
+		slog.F("template_id", ps.Preset.TemplateID.String()),
+		slog.F("template_name", ps.Preset.TemplateName),
+		slog.F("template_version_id", ps.Preset.TemplateVersionID),
+		slog.F("template_version_name", ps.Preset.TemplateVersionName),
+		slog.F("preset_id", ps.Preset.ID),
+		slog.F("preset_name", ps.Preset.Name),
+	)
+
+	actions, err := c.CalculateActions(ctx, ps)
+	if err != nil {
+		logger.Error(ctx, "failed to calculate actions for preset", slog.Error(err), slog.F("preset_id", ps.Preset.ID))
+		return nil
+	}
+
+	// nolint:gocritic // ReconcilePreset needs Prebuilds Orchestrator permissions.
+	prebuildsCtx := dbauthz.AsPrebuildsOrchestrator(ctx)
+
+	levelFn := logger.Debug
+	switch actions.ActionType {
+	case prebuilds.ActionTypeBackoff:
+		levelFn = logger.Warn
+	case prebuilds.ActionTypeCreate, prebuilds.ActionTypeDelete:
+		// Log at info level when there's a change to be effected.
+		levelFn = logger.Info
+	}
+
+	fields := []any{
+		slog.F("action_type", actions.ActionType),
+		slog.F("create_count", actions.Create),
+		slog.F("delete_count", len(actions.DeleteIDs)),
+		slog.F("to_delete", actions.DeleteIDs),
+	}
+	levelFn(ctx, "reconciliation actions for preset are calculated", fields...)
+
+	switch actions.ActionType {
+	case prebuilds.ActionTypeBackoff:
+		// If there is anything to backoff for (usually a cycle of failed prebuilds), then log and bail out.
+		levelFn(ctx, "template prebuild state retrieved, backing off",
+			append(fields,
+				slog.F("backoff_until", actions.BackoffUntil.Format(time.RFC3339)),
+				slog.F("backoff_secs", math.Round(actions.BackoffUntil.Sub(c.clock.Now()).Seconds())),
+			)...)
+
+		// return ErrBackoff
+		return nil
+
+	case prebuilds.ActionTypeCreate:
+		var multiErr multierror.Error
+
+		for range actions.Create {
```

Comment:
I've noticed that you removed this check: https://github.com/coder/coder/blob/d89f9543d380bddf15ee8d158521c3155dcaf4a5/enterprise/coderd/prebuilds/reconcile.go#L333-L341

What that intentional?

Reply:
@dannykopping  Initially I removed it - because code structure is changed.

But I put it back: https://github.com/coder/coder/pull/17261/commits/742d0d3046fa4da663083fe5da27dd35c9c54bbe